### ISSUE TITLE
🔥 Phase 2: GRPO Training Execution & Memory Profiling [PDE-21], LEG-4

### ISSUE THREAD

**Issue by clduab11** [clduab11/judicAIta#19]
## Objective

Execute a controlled 50-step GRPO training checkpoint to validate reward function composition, memory profiling, and estimate full training time within Kaggle's 9-hour constraint.

---

## Pre-Requisites

- [X] **LEG-3** (Phase 1) complete - all smoke tests passed
- [ ] **PR #7** merged (ground_truth metadata fix) - **CRITICAL BLOCKER**
- [ ] TPU runtime stable and memory visible

---

## 🔴 BLOCKER: PR #7 - ground_truth Metadata Bug

Before proceeding, verify this pattern exists in the training data:

```python
# Check for ground_truth field presence
import json
for sample in dataset['train'][:5]:
    if 'ground_truth' not in sample:
        raise ValueError("PR #7 NOT MERGED: ground_truth metadata missing!")
    print(f"✅ ground_truth: {sample['ground_truth'][:50]}...")
```

**If PR #7 is not merged, training will fail silently with incorrect reward signals.**

---

### 1\. Dataset Validation

- [ ] Load legal reasoning dataset from repo
- [ ] Verify `ground_truth` field present in all samples
- [ ] Confirm XML-tagged format: `<reasoning>...</reasoning><answer>...</answer>`
- [ ] Check sample distribution (min 100 samples for 50-step test)

### 2\. GRPO Configuration (50-Step Checkpoint)

```python
# Reduced configuration for validation run
grpo_config = {
    'num_generations': 2,          # Responses per iteration
    'num_iterations': 1,           # Policy optimization steps
    'beta': 0.08,                  # KL divergence penalty
    'epsilon': 0.2,                # PPO-like clipping
    'max_steps': 50,               # Checkpoint limit
}

optimizer_config = {
    'opt_type': 'adamw',
    'peak_value': 3e-6,
    'warmup_ratio': 0.1,
    'b1': 0.9,
    'b2': 0.999,
    'weight_decay': 0.01,
    'max_grad_norm': 1.0,
}

# Mesh configuration for TPU v3-8
mesh_config = {
    'shape': '(2,4)',
    'axis_names': "('fsdp','tp')",
}
```

### 3\. Reward Function Validation

Multi-component reward function (from Notion spec):

| Component | Weight | Validation Check |
| -- | -- | -- |
| Correctness | 40% | Answer matches ground_truth |
| Reasoning Quality | 30% | XML structure valid, logical flow |
| Citation Accuracy | 20% | Citations present and formatted |
| Clarity | 10% | Plain-English summary quality |

```python
def validate_reward_output(reward_dict):
    """Verify reward function returns expected structure"""
    assert 'correctness' in reward_dict
    assert 'reasoning_quality' in reward_dict
    assert 'citation_accuracy' in reward_dict
    assert 'clarity' in reward_dict
    assert sum(reward_dict.values()) <= 1.0  # Normalized
    return True
```

### 4\. Memory Profiling

```python
import jax

# Enable JAX profiler
jax.profiler.start_trace('/content/profiler_output')

# Run 10 training steps
for step in range(10):
    train_step(batch)
    
    # Log HBM usage every step
    from tunix.sft.utils import show_hbm_usage
    hbm_stats = show_hbm_usage()
    print(f"Step {step}: HBM used = {hbm_stats['used_gb']:.2f} GB")

jax.profiler.stop_trace()
```

**Memory Targets:**

* Peak HBM: < 12 GB (v3-8 has 16 GB per core)
* Gradient accumulation: Stable across steps
* No OOM errors through 50 steps

### 5\. Training Time Estimation

```python
import time

start_time = time.time()
# Run 50-step checkpoint
for step in range(50):
    train_step(batch)

elapsed = time.time() - start_time
time_per_step = elapsed / 50

# Extrapolate to full training
total_steps = estimate_total_steps(dataset_size, batch_size, epochs)
estimated_hours = (total_steps * time_per_step) / 3600

print(f"⏱️ 50 steps: {elapsed:.1f}s ({time_per_step:.2f}s/step)")
print(f"📊 Estimated full training: {estimated_hours:.1f} hours")

if estimated_hours > 8.5:
    print("⚠️ WARNING: Exceeds 9-hour Kaggle limit - reduce epochs or batch size")
```

---

## Code Wiki Cross-References

| Component | Tunix Path | Purpose |
| -- | -- | -- |
| GRPOLearner | `tunix/rl/grpo/grpo_learner.py` | Main GRPO training loop |
| RLCluster | `tunix/rl/rl_cluster.py` | Model loading, sharding, synchronization |
| PerfTracer | `tunix/perf/trace.py` | Host/device span capture |
| MetricsLogger | `tunix/sft/metrics_logger.py` | TensorBoard/W&B backends |
| Profiler | `tunix/sft/profiler.py` | JAX profiler activation |
| TFLOPS Calculator | `tunix/sft/system_metrics_calculator.py` | `measure_tflops_per_step()` |

**CLI Entry Point:**

```bash
python3 -m tunix.cli.grpo_main \
  base_config.yaml \
  reference_model_config.model_name="gemma-3-1b-it" \
  reference_model_config.mesh.shape="(2,4)" \
  rl_training_config.max_steps=50 \
  grpo_config.num_generations=2 \
  grpo_config.beta=0.08
```

---

## Hardware Resource Requirements (from Code Wiki)

| Model | Type | Min Resources | Max Batch Size | Sharding |
| -- | -- | -- | -- | -- |
| Gemma2 2B | Full/LoRA | v5e-4 | (4, 1) | Train: fsdp, Rollout: tp |
| Gemma 7B | Full/LoRA | v5p-2 | (4, 8) | Train: fsdp, Rollout: tp |

**Kaggle TPU v3-8 Considerations:**

* 8 cores, 16 GB HBM per core
* FSDP for training parallelism
* TP for rollout generation

---

## GitHub PR Context

| PR | Status | Impact on Phase 2 |
| -- | -- | -- |
| **PR #7** | 🔴 Open | **BLOCKER** - ground_truth metadata required for reward calculation |
| **PR #13** | ✅ Merged | Dependencies resolved |

---

## Validation Criteria

| Metric | Target | Blocker? |
| -- | -- | -- |
| 50-step completion | No OOM/errors | Yes |
| Reward signal | All 4 components returning values | Yes |
| Peak HBM | < 12 GB | Yes |
| Estimated time | < 8.5 hours | Yes |
| Gradient stability | No NaN/Inf | Yes |

---

## Success Criteria

✅ 50 training steps complete without error
✅ Reward function returns all 4 components correctly
✅ Memory stable under 12 GB peak
✅ Estimated full training < 8.5 hours
✅ JAX profiler trace saved for analysis

**→ Phase 3 can proceed when all checks pass**

**Comment by clduab11** [comment]
@clduab11  , ## Observations

The codebase implements GRPO training using **PyTorch** (file:src/judicaita/training/grpo_trainer.py), while the ticket references **Tunix/JAX/TPU** infrastructure that doesn't exist in the current codebase. The reward function has 3 components (Format, Outcome, Verbosity) rather than the 4 mentioned in the ticket (Correctness, Reasoning Quality, Citation Accuracy, Clarity). The `ground_truth` metadata field is referenced in scripts (file:scripts/update_notebook.py, file:scripts/update_notebook_data.py) but needs validation in the dataset loading pipeline. No memory profiling or TPU-specific utilities currently exist.

## Approach

The plan adapts the existing PyTorch GRPO implementation for a controlled 50-step validation run with memory profiling and time estimation. Since the codebase uses PyTorch/CUDA rather than JAX/TPU, the implementation will focus on GPU memory tracking (CUDA memory instead of HBM), gradient stability monitoring, and training time estimation. The reward function will be enhanced to match the 4-component specification from the ticket, and dataset validation will ensure `ground_truth` field presence before training begins.

## Implementation Steps

### 1. Dataset Validation & Ground Truth Verification

**Objective**: Ensure all training samples contain the `ground_truth` field required for reward calculation.

**Files to Modify**:
- file:src/judicaita/training/data_curation.py

**Implementation**:
- Add validation function `validate_ground_truth_presence()` in `LegalBenchDataset.load()` method
- Check each sample for `ground_truth` field after dataset loading
- Raise `ValueError` with clear message if field is missing (indicating PR #7 not merged)
- Add logging to show sample count and ground_truth field statistics
- Verify XML-tagged format: `<reasoning>...</reasoning><answer>...</answer>` in responses

**Validation Check**:
```python
# In LegalBenchDataset.load() after dataset creation
for i, sample in enumerate(self.dataset):
    if 'ground_truth' not in sample:
        raise ValueError(f"PR #7 NOT MERGED: ground_truth missing in sample {i}")
    if not sample['ground_truth']:
        logger.warning(f"Empty ground_truth in sample {i}")
```

---

### 2. Enhanced Reward Function with 4 Components

**Objective**: Align reward function with ticket specification (Correctness 40%, Reasoning Quality 30%, Citation Accuracy 20%, Clarity 10%).

**Files to Modify**:
- file:src/judicaita/training/rewards.py

**Implementation**:
- Rename `OutcomeReward` to `CorrectnessReward` (40% weight) - validates answer matches ground_truth
- Rename `FormatReward` to `ReasoningQualityReward` (30% weight) - validates XML structure and logical flow
- Create new `CitationAccuracyReward` class (20% weight) - checks for legal citations (U.S.C., case names, section symbols)
- Create new `ClarityReward` class (10% weight) - evaluates plain-English quality using readability metrics
- Update `CompositeReward` to use new weights: `correctness_weight=0.4, reasoning_quality_weight=0.3, citation_accuracy_weight=0.2, clarity_weight=0.1`
- Add validation method `validate_reward_output()` to ensure all 4 components return values and sum ≤ 1.0

**Component Breakdown**:

| Component | Weight | Implementation |
|-----------|--------|----------------|
| Correctness | 40% | Token overlap or exact match with ground_truth |
| Reasoning Quality | 30% | XML tag validation + sentence coherence |
| Citation Accuracy | 20% | Regex patterns for legal citations (U.S.C., v., §) |
| Clarity | 10% | Flesch reading ease or sentence length metrics |

---

### 3. Memory Profiling Infrastructure

**Objective**: Track GPU memory usage throughout training to ensure peak usage stays under limits.

**Files to Create**:
- file:src/judicaita/training/profiler.py

**Files to Modify**:
- file:src/judicaita/training/grpo_trainer.py

**Implementation**:
- Create `MemoryProfiler` class with methods:
  - `get_gpu_memory_stats()` - returns dict with `allocated_gb`, `reserved_gb`, `max_allocated_gb`
  - `log_memory_usage(step)` - logs memory at specific training step
  - `check_memory_threshold(threshold_gb)` - raises warning if exceeded
- Integrate into `GRPOTrainer._grpo_step()` to track memory every N steps
- Add memory logging to `TrainingConfig` with `memory_log_steps` parameter (default: 10)
- Store memory statistics in training metrics dict

**Memory Tracking**:
```python
# In GRPOTrainer.train() loop
if global_step % self.config.memory_log_steps == 0:
    memory_stats = self.profiler.get_gpu_memory_stats()
    logger.info(f"Step {global_step}: GPU Memory = {memory_stats['allocated_gb']:.2f} GB")
    training_metrics['memory_usage'].append(memory_stats)
```

---

### 4. 50-Step Checkpoint Configuration

**Objective**: Create configuration for controlled 50-step validation run.

**Files to Modify**:
- file:src/judicaita/core/config.py
- file:src/judicaita/cli.py

**Implementation**:
- Add `max_steps` parameter to `TrainingConfig` (default: None for full training)
- Add `checkpoint_validation_mode` boolean flag
- Modify `GRPOTrainer.train()` to respect `max_steps` limit and exit early
- Add CLI flag `--max-steps` to `train_grpo` command
- Create validation config preset with reduced parameters:
  - `num_rollouts=2` (vs 4 for full training)
  - `batch_size=4`
  - `gradient_accumulation_steps=4`
  - `max_steps=50`

**CLI Usage**:
```bash
```

---

### 5. Training Time Estimation

**Objective**: Measure time per step and extrapolate to full training duration.

**Files to Modify**:
- file:src/judicaita/training/grpo_trainer.py

**Implementation**:
- Add `time_per_step` tracking in `GRPOTrainer.train()` using `time.time()`
- Calculate average time per step over 50-step run
- Compute estimated total training time: `(total_steps * avg_time_per_step) / 3600`
- Log warning if estimated time exceeds 8.5 hours (Kaggle limit with buffer)
- Store timing metrics in training results dict

**Estimation Logic**:
```python
# After 50-step checkpoint
elapsed = time.time() - start_time
time_per_step = elapsed / 50
total_steps = (len(dataset) // batch_size // grad_accum_steps) * num_epochs
estimated_hours = (total_steps * time_per_step) / 3600

if estimated_hours > 8.5:
    logger.warning(f"Estimated {estimated_hours:.1f}h exceeds 9h limit - reduce epochs/batch")
```

---

### 6. Gradient Stability Monitoring

**Objective**: Detect NaN/Inf gradients that indicate training instability.

**Files to Modify**:
- file:src/judicaita/training/grpo_trainer.py

**Implementation**:
- Add gradient norm tracking in `_grpo_step()` before clipping
- Check for NaN/Inf values using `torch.isnan()` and `torch.isinf()`
- Log gradient statistics (mean, max, min) every logging step
- Raise error and save emergency checkpoint if NaN/Inf detected
- Store gradient norms in training metrics

**Gradient Check**:
```python
# After loss.backward() and before optimizer.step()
total_norm = torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.config.max_grad_norm)
if torch.isnan(total_norm) or torch.isinf(total_norm):
    logger.error(f"NaN/Inf gradient at step {global_step}")
    self.save_checkpoint(f"emergency-step-{global_step}")
    raise RuntimeError("Training unstable: NaN/Inf gradients detected")
training_metrics['gradient_norms'].append(total_norm.item())
```

---

### 7. Validation Criteria Checklist

**Objective**: Automated validation of all success criteria from ticket.

**Files to Create**:
- file:src/judicaita/training/validation.py

**Implementation**:
- Create `ValidationChecker` class with methods for each criterion:
  - `check_completion(metrics)` - verifies 50 steps completed without OOM
  - `check_reward_components(reward_results)` - ensures all 4 components return values
  - `check_memory_limit(memory_stats, threshold=12.0)` - validates peak memory < threshold
  - `check_time_estimate(estimated_hours, limit=8.5)` - validates training time feasible
  - `check_gradient_stability(gradient_norms)` - ensures no NaN/Inf
- Generate validation report at end of 50-step run
- Return boolean pass/fail for each criterion

**Validation Report**:
```python
validation_report = {
    'completion': True,
    'reward_components': True,
    'memory_stable': True,
    'time_feasible': True,
    'gradients_stable': True,
    'ready_for_phase_3': all([...])
}
```

---

### 8. CLI Integration & Execution

**Objective**: Provide simple CLI command for Phase 2 validation run.

**Files to Modify**:
- file:src/judicaita/cli.py

**Implementation**:
- Add `--validation-mode` flag to `train_grpo` command
- When enabled, automatically sets: `max_steps=50`, `num_rollouts=2`, enables memory profiling
- Add `--memory-threshold` parameter (default: 12.0 GB for GPU, adjust for hardware)
- Add `--time-limit` parameter (default: 8.5 hours)
- Print validation report at completion
- Exit with code 0 if all checks pass, 1 if any fail

**Example Command**:
```bash
```

---

## Execution Flow

```mermaid
sequenceDiagram
    participant CLI as CLI Command
    participant Validator as Dataset Validator
    participant Trainer as GRPO Trainer
    participant Profiler as Memory Profiler
    participant Reward as Reward Function
    participant Checker as Validation Checker

    CLI->>Validator: Load dataset
    Validator->>Validator: Check ground_truth fields
    alt ground_truth missing
        Validator-->>CLI: Raise ValueError (PR #7 blocker)
    end
    
    CLI->>Trainer: Initialize with 50-step config
    Trainer->>Profiler: Start memory tracking
    
    loop 50 training steps
        Trainer->>Reward: Compute 4-component rewards
        Reward-->>Trainer: Return reward scores
        Trainer->>Trainer: Backward pass
        Trainer->>Profiler: Log memory usage
        Trainer->>Trainer: Check gradient stability
    end
    
    Trainer->>Checker: Validate completion criteria
    Checker->>Checker: Check memory, time, gradients
    Checker-->>CLI: Return validation report
    
    alt All checks pass
        CLI->>CLI: Print success + Phase 3 ready
    else Any check fails
        CLI->>CLI: Print failure + recommendations
    end
```

---

## Success Criteria Mapping

| Criterion | Implementation | Validation Method |
|-----------|----------------|-------------------|
| 50-step completion | `max_steps=50` in TrainingConfig | Check `global_step == 50` without exceptions |
| Reward 4 components | Enhanced CompositeReward | Validate all keys present in reward_result.details |
| Peak memory < 12 GB | MemoryProfiler tracking | Check `max(memory_stats['allocated_gb']) < 12.0` |
| Estimated time < 8.5h | Time per step measurement | Calculate `total_steps * avg_time / 3600 < 8.5` |
| Gradient stability | NaN/Inf detection | Verify no NaN/Inf in `gradient_norms` list |

---

## Hardware Considerations

**Note**: The ticket mentions TPU v3-8 with HBM tracking, but the codebase uses PyTorch/CUDA. The implementation assumes GPU training:

- **GPU Memory**: Track using `torch.cuda.memory_allocated()` and `torch.cuda.max_memory_allocated()`
- **Target**: < 12 GB peak (adjust based on available GPU memory)
- **Fallback**: If TPU training is required, a separate JAX/Tunix implementation would be needed (not in current codebase)

**For TPU Adaptation** (if needed):
- Would require rewriting training loop in JAX/Flax
- Use `jax.profiler` instead of PyTorch profilers
- Track HBM using TPU-specific APIs
- Reference file:docs/GRPO_TRAINING.md mentions TPU support but not implemented

---

## Post-Validation Actions

After successful 50-step validation:

1. **Review Metrics**: Analyze memory profile, reward distributions, gradient norms
2. **Adjust Hyperparameters**: If time estimate too high, reduce epochs or increase batch size
3. **Full Training**: Remove `max_steps` limit and run complete training
4. **Phase 3 Readiness**: Confirm all blockers resolved before proceeding

**If Validation Fails**:
- Memory exceeded → Reduce batch size or enable gradient checkpointing
- Time too long → Reduce num_rollouts or use smaller model
- Gradients unstable → Lower learning rate or adjust GRPO tau
- Reward components missing → Fix reward function implementation
